### PR TITLE
Download slack files using API

### LINKED ISSUE #141: Download slack files using API
My requirement is to download files from slack. Is there any method to download the files. I could not find any. Thanks!

### PR DESCRIPTION
**Pull request by jmswenski**
Adding download functionality to the client with a new method GetFile
which will write the contents to a provided io.Writer

Resolves: #141

### CODE REVIEW

**Comment by jmswenski** [comment]
Open to feedback - just started messing around with go the other day so if its horrible, will gladly make changes.

File: files.go [hunk 1]
```diff
@@ -153,6 +153,12 @@ func (api *Client) GetFileInfo(fileID string, count, page int) (*File, []Comment
 	return api.GetFileInfoContext(context.Background(), fileID, count, page)
 }
 
+//GetFile downloads a file
+func (api *Client) GetFile(file *File, writer io.Writer) error {
+	err := getFile(context.Background(), api.httpclient, file.URLPrivateDownload, api.token, writer)
```
Comment (james-lawrence) [inline files.go:158]: just straight return here.

File: misc.go [hunk 2]
```diff
@@ -37,6 +37,45 @@ func (e *RateLimitedError) Error() string {
...
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusTooManyRequests {
```
Comment (james-lawrence) [inline misc.go:60]: maybe create a helper method for handling errors, this code is littered all over the place.

**Review by james-lawrence — COMMENTED** [review]
overall looks good, any chance you could write a test, so we don't break the method going forward?

File: misc.go [hunk 3]
```diff
@@ -37,6 +37,45 @@ func (e *RateLimitedError) Error() string {
 	return fmt.Sprintf("Slack rate limit exceeded, retry after %s", e.RetryAfter)
 }
 
+func getFile(ctx context.Context, client HTTPRequester, url, token string, writer io.Writer) error {
+	body := &bytes.Buffer{}
+
+	req, err := http.NewRequest("GET", url, body)
```
Comment (james-lawrence) [inline misc.go:43]: this will likely segfault if NewRequest actually errors out, as req will be nil.

**Comment by james-lawrence** [comment]
simple test and fixing the seg fault and this can be merged if anyone is interested in taking up the work.

**Comment by alyosha** [comment]
picked up this task in a new PR 
https://github.com/nlopes/slack/pull/472

**Comment by alyosha** [comment]
think we can probably close this since #472 resolved the issue?

**Comment by alyosha** [comment]
or at least maybe take off the help wanted labels to prevent confusion

**Comment by james-lawrence** [comment]
closing due to age.

**Pull request closed without merging**